src/fastcs/transport/epics/util.py

```diff
@@ -25,75 +31,104 @@
-
-    Returns:
-        `True` if `Attribute` is an enum, else `False`
-
-    """
-    match attribute:
-        case Attribute(datatype=String(), allowed_values=allowed_values) if (
-            allowed_values is not None and len(allowed_values) <= MBB_MAX_CHOICES
-        ):
-            return True
+EPICS_ALLOWED_DATATYPES = (Bool, DataType, Enum, Float, Int, String, WaveForm)
+
+DATATYPE_FIELD_TO_RECORD_FIELD = {
+    "prec": "PREC",
+    "units": "EGU",
+    "min": "DRVL",
+    "max": "DRVH",
+    "min_alarm": "LOPR",
+    "max_alarm": "HOPR",
+}
+
+
+def record_metadata_from_attribute(
+    attribute: Attribute[T],
+) -> dict[str, str | None]:
+    return {"DESC": attribute.description}
+
+
+def record_metadata_from_datatype(datatype: DataType[T]) -> dict[str, str]:
+    arguments = {
+        DATATYPE_FIELD_TO_RECORD_FIELD[field]: value
+        for field, value in asdict(datatype).items()
+        if field in DATATYPE_FIELD_TO_RECORD_FIELD
+    }
+
+    match datatype:
+        case WaveForm():
+            if len(datatype.shape) != 1:
+                raise TypeError(
+                    f"Unsupported shape {datatype.shape}, the EPICS backend only "
+                    "supports to 1D arrays"
+                )
+            arguments["length"] = datatype.shape[0]
+        case Enum():
+            if len(datatype.members) <= MBB_MAX_CHOICES:
+                state_keys = dict(
+                    zip(
+                        MBB_STATE_FIELDS,
+                        [member.name for member in datatype.members],
+                        strict=False,
+                    )
+                )
+                arguments.update(state_keys)
+
+    return arguments
+
+
+def get_callable_from_epics_type(datatype: DataType[T]) -> Callable[[object], T]:
+    match datatype:
+        case Enum():
+
+            def cast_from_epics_type(value: object) -> T:
+                return datatype.validate(datatype.members[value])
+
+        case datatype if issubclass(type(datatype), EPICS_ALLOWED_DATATYPES):
+
+            def cast_from_epics_type(value) -> T:
+                return datatype.validate(value)
         case _:
-            return False
-
-
-def enum_value_to_index(attribute: Attribute[T], value: T) -> int:
-    """Convert the given value to the index within the allowed_values of the Attribute
-
-    Args:
-        `attribute`: The attribute
-        `value`: The value to convert
-
-    Returns:
-        The index of the `value`
+            raise ValueError(f"Unsupported datatype {datatype}")
+    return cast_from_epics_type
 
-    Raises:
-        ValueError: If `attribute` has no allowed values or `value` is not a valid
-            option
 
-    """
-    if attribute.allowed_values is None:
-        raise ValueError(
-            "Cannot convert value to index for Attribute without allowed values"
-        )
+def get_callable_to_epics_type(datatype: DataType[T]) -> Callable[[T], object]:
```

Comment:
I think these methods would be simpler if we pass the datatype along with the value and always do the match on get/put.